gnomad_qc/v4/assessment/calculate_per_sample_stats.py

```diff
@@ -187,64 +188,64 @@ def compute_agg_sample_stats(
+         working on "exomes" data.
     :return: Struct of aggregate statistics for per-sample QC metrics.
     """
     if by_ancestry and meta_ht is None:
         raise ValueError(
             "If `by_ancestry` is True, a Table containing sample metadata is required."
         )
 
-    agg_expr = {
-        strat: hl.struct(
-            **{
-                metric: hl.struct(
-                    mean=hl.agg.mean(ht[strat][metric]),
-                    quantiles=hl.agg.approx_quantiles(
-                        ht[strat][metric], [0.0, 0.25, 0.5, 0.75, 1.0]
-                    ),
-                )
-                for metric in ht[strat]
-                if isinstance(ht[strat][metric], hl.expr.NumericExpression)
-            }
-        )
-        for strat in ht.row_value
-        if isinstance(ht[strat], hl.expr.StructExpression)
-    }
-
-    stats_struct = ht.aggregate(agg_expr, _localize=False)
-
-    if by_ancestry:
-        # Performing aggregation for all samples and by genetic ancestry.
-        # Note: This is performing multiple aggregations, rather than a single group_by
-        # aggregation to avoid a Class too large error.
-        gen_anc_expr = meta_ht.population_inference.pop
-        gen_ancs = meta_ht.aggregate(
-            hl.agg.filter(
-                hl.is_defined(gen_anc_expr), hl.agg.collect_as_set(gen_anc_expr)
-            )
-        )
-        gen_anc_expr = meta_ht[ht.s].population_inference.pop
+    subset = (
+        ["gnomad"]
+        if not by_subset
+        else [
+            "gnomad",
+            hl.if_else(meta_ht[ht.s].project_meta.ukb_sample, "ukb", "non-ukb"),
+        ]
+    )
+    gen_anc = (
+        ["global"]
+        if not by_ancestry
+        else ["global", meta_ht[ht.s].population_inference.pop]
+    )
 
-        stats_struct = hl.struct(
-            all_samples=stats_struct,
-            **{
-                anc: ht.aggregate(
-                    hl.agg.filter(gen_anc_expr == anc, agg_expr), _localize=False
-                )
-                for anc in gen_ancs
-            },
-        )
+    all_strats = [
+        s for s in ht.row_value if isinstance(ht[s], hl.expr.StructExpression)
+    ]
+
+    ht = ht.transmute(
+        subset=subset,
+        gen_anc=gen_anc,
+        _stats_array=[(s, ht[s]) for s in all_strats],
+    )
+
+    ht = ht.explode("_stats_array").explode("gen_anc").explode("subset")
 
-    return stats_struct
+    ht = ht.group_by("subset", "gen_anc", variant_filter=ht._stats_array[0]).aggregate(
+        **{
+            m: hl.struct(
+                mean=hl.agg.mean(ht._stats_array[1][m]),
+                quantiles=hl.agg.approx_quantiles(
+                    ht._stats_array[1][m], [0.0, 0.25, 0.5, 0.75, 1.0]
+                ),
+            )
+            for m in ht._stats_array[1]
+            if isinstance(ht._stats_array[1][m], hl.expr.NumericExpression)
+        }
+    )
+
+    return ht
```

Comment:
is this Julia's new efficient code? Let me take some more time today to make sure I understand it, I'll lyk any questions tomorrow

Reply:
Yes, I adapted hers for by_ancestry and by_subset.